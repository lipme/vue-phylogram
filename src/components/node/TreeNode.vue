// Copyright 2020 [LIPM]
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

// http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
//    limitations under the License.
<template>
<g :transform="transformAttribute">
    <circle v-if="!collapsed" :class="{transparent: !show}" :r="size/2"
     :stroke-width="strokeWidth" :stroke="strokeColor" :fill="fill"
     >
      <animateTransform attributeName="transform"
             type="scale"
             from="0 0"
             to="1 1"
             begin="0s"
             dur="1s"
        />
     </circle>
       <polygon v-else :points="triangleCoords" :fill="fill" :stroke-width="strokeWidth" :stroke="strokeColor"/>
    </g>
</template>

<script>
export default {
  name: 'TreeNode',
  props: {
    size: {
      type: Number,
      default: 6
    },
    x: {
      type: Number,
      default: 0
    },
    y: {
      type: Number,
      default: 0
    },
    circular: {
      type: Boolean,
      default: false
    },
    show: {
      type: Boolean,
      default: true
    },
    strokeColor: {
      type: String,
      default: 'darkblue'
    },
    fill: {
      type: String,
      default: 'steelblue'
    },
    collapsed: {
      type: Boolean,
      default: false
    }
  },

  computed: {
    strokeWidth () {
      return this.size / 5
    },
    transformAttribute () {
      if (this.circular === false) {
        return 'translate(' + this.y + ' ' + this.x + ')'
      } else {
        return 'rotate(' + (this.x - 90) + ')translate(' + this.y + ')'
      }
    },
    triangleCoords () {
      const coords = [
        { x: 0, y: 0 },
        {
          x: this.size * 1.5,
          y: -this.size * 1.5
        },
        {
          x: this.size * 1.5,
          y: this.size * 1.5
        }
      ]
      return `${coords[0].x} ${coords[0].y},${coords[1].x} ${coords[1].y}, ${coords[2].x} ${coords[2].y}`
    }
  }
}
</script>

<style scoped>

.transparent {
  opacity:0;
}

circle:hover,polygon:hover {
  transition: all 0.5s;
  transform: scale(2);
  opacity: 1;
}

</style>
