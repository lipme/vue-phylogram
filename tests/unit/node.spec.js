// Copyright 2020 [LIPM]
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

// http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
//    limitations under the License.
import { shallowMount } from '@vue/test-utils'
import TreeNode from '@/components/node'

describe('TreeNode component', () => {
  it('default node renders correctly', () => {
    const comp = shallowMount(TreeNode)

    expect(comp.element).toMatchSnapshot()
  })
  it('node size set correctly', () => {
    const comp = shallowMount(TreeNode, {
      propsData: {
        size: 10
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('node x set correctly', () => {
    const comp = shallowMount(TreeNode, {
      propsData: {
        x: 100
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('node y set correctly', () => {
    const comp = shallowMount(TreeNode, {
      propsData: {
        y: 100
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('node position in circular mode set correctly', () => {
    const comp = shallowMount(TreeNode, {
      propsData: {
        circular: true
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('show mode set correctly', () => {
    const comp = shallowMount(TreeNode, {
      propsData: {
        show: false
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('stroke color set correctly', () => {
    const comp = shallowMount(TreeNode, {
      propsData: {
        strokeColor: 'black'
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('fill color set correctly', () => {
    const comp = shallowMount(TreeNode, {
      propsData: {
        fill: 'black'
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('collapsed mode set correctly', () => {
    const comp = shallowMount(TreeNode, {
      propsData: {
        collapsed: true
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
})
