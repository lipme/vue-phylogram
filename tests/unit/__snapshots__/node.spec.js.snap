// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeNode component collapsed mode set correctly 1`] = `
<g
  transform="translate(0 0)"
>
  <polygon
    fill="steelblue"
    points="0 0,9 -9, 9 9"
    stroke="darkblue"
    stroke-width="1.2"
  />
</g>
`;

exports[`TreeNode component default node renders correctly 1`] = `
<g
  transform="translate(0 0)"
>
  <circle
    class=""
    fill="steelblue"
    r="3"
    stroke="darkblue"
    stroke-width="1.2"
  >
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </circle>
</g>
`;

exports[`TreeNode component fill color set correctly 1`] = `
<g
  transform="translate(0 0)"
>
  <circle
    class=""
    fill="black"
    r="3"
    stroke="darkblue"
    stroke-width="1.2"
  >
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </circle>
</g>
`;

exports[`TreeNode component node position in circular mode set correctly 1`] = `
<g
  transform="rotate(-90)translate(0)"
>
  <circle
    class=""
    fill="steelblue"
    r="3"
    stroke="darkblue"
    stroke-width="1.2"
  >
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </circle>
</g>
`;

exports[`TreeNode component node size set correctly 1`] = `
<g
  transform="translate(0 0)"
>
  <circle
    class=""
    fill="steelblue"
    r="5"
    stroke="darkblue"
    stroke-width="2"
  >
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </circle>
</g>
`;

exports[`TreeNode component node x set correctly 1`] = `
<g
  transform="translate(0 100)"
>
  <circle
    class=""
    fill="steelblue"
    r="3"
    stroke="darkblue"
    stroke-width="1.2"
  >
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </circle>
</g>
`;

exports[`TreeNode component node y set correctly 1`] = `
<g
  transform="translate(100 0)"
>
  <circle
    class=""
    fill="steelblue"
    r="3"
    stroke="darkblue"
    stroke-width="1.2"
  >
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </circle>
</g>
`;

exports[`TreeNode component show mode set correctly 1`] = `
<g
  transform="translate(0 0)"
>
  <circle
    class="transparent"
    fill="steelblue"
    r="3"
    stroke="darkblue"
    stroke-width="1.2"
  >
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </circle>
</g>
`;

exports[`TreeNode component stroke color set correctly 1`] = `
<g
  transform="translate(0 0)"
>
  <circle
    class=""
    fill="steelblue"
    r="3"
    stroke="black"
    stroke-width="1.2"
  >
    <animatetransform
      attributeName="transform"
      begin="0s"
      dur="1s"
      from="0 0"
      to="1 1"
      type="scale"
    />
  </circle>
</g>
`;
